src/python/pants/backend/python/pip_requirement.py

```diff
@@ -21,8 +22,35 @@ class PipRequirement:
     """
 
     @classmethod
-    def parse(cls, line: str) -> PipRequirement:
-        return cls(pkg_resources.Requirement.parse(line))
+    def parse(cls, line: str, description_of_origin: str = "") -> PipRequirement:
+        try:
+            return cls(pkg_resources.Requirement.parse(line))
+        except Exception as e:
+            scheme, netloc, path, query, fragment = urllib.parse.urlsplit(line)
+            if scheme and fragment.startswith("egg="):
+                # Try converting a pip VCS-style requirement into a PEP-440 one that can be
+                # parsed as a Requirement. E.g.,
+                # git+https://github.com/django/django.git@stable/2.1.x#egg=Django
+                # into
+                # Django@ git+https://github.com/django/django.git@stable/2.1.x
+                pep_440_vcs_url = urllib.parse.urlunsplit((scheme, netloc, path, query, ""))
+                project = fragment[4:]
+                pep_440_req_str = f"{project}@ {pep_440_vcs_url}"
+                try:
+                    ret = cls(pkg_resources.Requirement.parse(pep_440_req_str))
+                    logger.warning(
+                        "Converted a pip VCS requirement to a PEP440-compatible one:\n"
```

Comment:
This could just be a `file:///here#egg=...` and the #egg is even optional (or even a `/here` or `.` requirement).

Reply:
If #egg is not present then I think we have to fail?

Reply:
Pex and Pip do not fail for these. It may be the case that Pants wants to continue to restrict to PEP-440 though. Probably more a comment to prompt thinking about, what else are we continuing to cut off because we're expressing an opinion about a string. I actually don't know the answer of why we try parse these at all - is it just to get a project name for dep inference?